File: 02.IntroductionToAlgorithmsAndDataStructures/alds1-07-b.Binary-Tree.rb
class Node
    attr_accessor :parent, :left, :right
    
    def initialize(parent, left, right)
        @parent = parent
        @left   = left
        @right  = right
    end
end

def sibling(key, nodes)
    p = nodes[key].parent
    if p != -1
        if nodes[p].left == key
            nodes[p].right
        else
            nodes[p].left
        end
    else
        -1
    end
end

def degree(key, nodes)
    if nodes[key].left != -1 && nodes[key].right != -1
        2
    elsif nodes[key].left != -1 || nodes[key].right != -1
        1
    else
        0
    end
end

def depth(key, nodes)
    depth = 0
    i = key
    while nodes[i].parent != -1
        depth += 1
        i = nodes[i].parent
    end
    depth
end

def height(key, nodes)
    left_h = right_h = 0
    if nodes[key].left != -1
        left_h = 1 + height(nodes[key].left, nodes)
    end

    if nodes[key].right != -1
        right_h = 1 + height(nodes[key].right, nodes)
    end
    left_h > right_h ? left_h : right_h
end

def kind(key, nodes)
    if nodes[key].parent == -1
        "root"
    elsif nodes[key].left == -1 && nodes[key].right == -1
        "leaf"
    else
        "internal node"
    end
end

nodes = {}
n = gets.chomp.to_i
n.times{
    k, l, r = gets.chomp.split.map(&:to_i)
    if nodes[k] != nil
        nodes[k].left = l
        nodes[k].right = r
    else
        nodes[k] = Node.new(-1, l, r)
    end

    #-- left格納
    if l != -1 && nodes[l] != nil
        nodes[l].parent = k
    elsif l != -1
        nodes[l] = Node.new(k, -1 , -1)
    end
    
    #-- right格納
    if r != -1 && nodes[r] != nil
        nodes[r].parent = k
    elsif r != -1
        nodes[r] = Node.new(k, -1, -1)
    end
}

[*0...n].each do |i|
    print "node #{i}: parent = #{nodes[i].parent}"
    print ", sibling = #{sibling(i, nodes)}"
    print ", degree = #{degree(i, nodes)}"
    print ", depth = #{depth(i, nodes)}"
    print ", height = #{height(i, nodes)}"
    puts  ", #{kind(i, nodes)}"
end

#-- require 'minitest/autorun'
#-- class TestBinTree < Minitest::Test
  #-- def test_height
    #-- require 'byebug'; byebug
    #-- assert_equal 3, height(0, @@nodes)
  #-- end
#-- end

File: 02.IntroductionToAlgorithmsAndDataStructures/alds1-07-b.Binary-Tree.hs
import Control.Monad
import Data.List
import Data.Ord
 
inputList 0 _ = []
inputList n (a:b:c:rest) =
    [a, b, c] : inputList (n-1) rest
 
parent [] n = -1
parent ([x, y, z]:xs) n =
    if y == n || z == n then x
    else parent xs n
     
findNode [] n = [-1, -1, -1] -- fxxking node
findNode ([x, y, z]:xs) n
    | x == n = [x, y, z]
    | otherwise = findNode xs n
 
sibling xs n = sibling_ xs n (parent xs n)
sibling_ xs n p
    | p == -1 = -1
    | otherwise = sibling__ xs n (findNode xs p)
sibling__ xs n [_, q, r]
    | q == n = r
    | otherwise = q
 
degree [x, y, z] = (zeroOrOne (y /= -1)) + (zeroOrOne (z /= -1))
    where zeroOrOne bl = if bl then 1 else 0
 
depth xs n = depth_ $ parent xs n
    where depth_ p = if p == -1 then 0 else (depth xs p) + 1
 
height xs (-1) = -1
height xs n = (max (height xs y) (height xs z)) + 1
    where [x, y, z] = findNode xs n
     
printType xs [x, y, z]
    | parent xs x == -1 = "root"
    | degree [x, y, z] == 0 = "leaf"
    | otherwise = "internal node"
 
printNode xs node =
    "node " ++ (show id) ++ ": parent = " ++ (show $ parent xs id)
    ++ ", sibling = " ++ (show $ sibling xs id)
    ++ ", degree = " ++ (show $ degree node)
    ++ ", depth = " ++ (show $ depth xs id)
    ++ ", height = " ++ (show $ height xs id)
    ++ ", " ++ (printType xs node)
    where [id, left, right] = node
 
main = do
    str <- getContents
    let xs1 = map (\x -> read x :: Int) (words str)
    let xs2 = inputList (head xs1) (tail xs1)
    let xs3 = sortBy (comparing head) xs2
    let res = foldl (\acc x -> acc ++ (printNode xs3 x) ++ "\n") "" xs3
    putStr res

